crates/fuel-core/src/graphql_api/worker_service.rs

```diff
@@ -181,16 +207,121 @@ where
     }
 }
 
+trait HasIndexation<'a> {
+    type Storage: Mappable;
+
+    fn key(&self) -> <Self::Storage as Mappable>::Key;
+    fn amount(&self) -> Amount;
+    fn update_balances<T, F>(&self, tx: &mut T, updater: F) -> StorageResult<()>
+    where
+        <<Self as HasIndexation<'a>>::Storage as Mappable>::Key:
+            Sized + std::fmt::Display,
+        <<Self as HasIndexation<'a>>::Storage as Mappable>::Value:
+            Sized + std::fmt::Display,
+        <<Self as HasIndexation<'a>>::Storage as Mappable>::OwnedValue:
+            Default + std::fmt::Display,
+        T: OffChainDatabaseTransaction
+            + StorageInspect<
+                <Self as HasIndexation<'a>>::Storage,
+                Error = fuel_core_storage::Error,
+            > + StorageMutate<<Self as HasIndexation<'a>>::Storage>,
+        F: Fn(
+            Cow<<<Self as HasIndexation<'a>>::Storage as Mappable>::OwnedValue>,
+            Amount,
+        ) -> <<Self as HasIndexation<'a>>::Storage as Mappable>::Value,
+        fuel_core_storage::Error:
+            From<<T as StorageInspect<<Self as HasIndexation<'a>>::Storage>>::Error>,
+    {
+        let key = self.key();
+        let amount = self.amount();
+        let storage = tx.storage::<Self::Storage>();
+        let current_balance = storage.get(&key)?.unwrap_or_default();
```

Comment:
Do we risk losing data here, e.g. in the case of parallel updates to the balance?

Reply:
All storage calls in this function come from the transaction that we [pass](https://github.com/FuelLabs/fuel-core/pull/2383/files#diff-1ac4adf5631afae925791c3897c52c2fa6282bd70901fddf552e7e1f0492c447R212) in.

This function can be invoked from two contexts (graphql worker, genesis importer), each providing a pre-existing DB transaction.